src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs

```diff
@@ -0,0 +1,184 @@
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    /// <summary>
+    /// Parse a project.json, and add additional dependencies, then write a out new project.json.
+    /// </summary>
+    public class AddDependenciesToProjectJson : Task
+    {
+        private Regex VersionStructureRegex = new Regex(@"(\d+\.\d+\.\d+)-((beta|rc2|rc3)?-?(\d+))");
+        private Regex BuildNumberOverrideStructureRegex = new Regex(@"(beta|rc2|rc3)?-?(\d+)?");
+        private Regex ThreePartVersionRegex = new Regex(@"(\d+\.\d+\.\d+)");
+
+        // Additional Dependencies to add to the project.json. May Optionally contain a version.
+        // Will Override dependencies present in the project if there is a conflict.
+        // AdditionalDependencies required metadata:  Name, Version
+
+        [Required]
+        public ITaskItem[] AdditionalDependencies { get; set; }
+
+        // Permit overriding package versions found in project.json with custom build number version.
+        public string PackageBuildNumberOverride { get; set; }
+
+        // Original package version which is used to seed the output project.json
+        [Required]
+        public String ProjectJson { get; set; }
+
+        // The directory to put the generated project.json in
+        [Required]
+        public String OutputProjectJson { get; set; }
+
+        public override bool Execute()
+        {
+            if (!File.Exists(ProjectJson))
+            {
+                Log.LogError("Cannot find specified project.json - '{0}'", ProjectJson);
+                return false;
+            }
+            if(!string.IsNullOrWhiteSpace(PackageBuildNumberOverride))
+            {
+                if(!BuildNumberOverrideStructureRegex.IsMatch(PackageBuildNumberOverride))
+                {
+                    Log.LogError("Invalid package version format: '{0}'", PackageBuildNumberOverride);
+                    return false;
+                }
+            }
+
+            JObject projectRoot = ReadProject(ProjectJson);
+            JObject dependencies = GenerateDependencies(projectRoot);
+            projectRoot = UpdateProperty("dependencies", projectRoot, dependencies);
+            WriteProject(projectRoot, OutputProjectJson);
+
+            return true;
+        }
+
+        private static JObject ReadProject(string projectJsonPath)
+        { 
+           using (TextReader projectFileReader = File.OpenText(projectJsonPath)) 
+           { 
+              var projectJsonReader = new JsonTextReader(projectFileReader); 
+              var serializer = new JsonSerializer(); 
+              return serializer.Deserialize<JObject>(projectJsonReader); 
+           } 
+        }
+
+        // Generate the combines dependencies from the projectjson jObject and from AdditionalDependencies
+        private JObject GenerateDependencies(JObject projectJsonRoot)
+        {
+            var originalDependenciesList = new List<JToken>();
```

Comment:
I think we could factor out a `virtual VisitDependencies` method in `VisitProjectDependencies` and then this would fit in as a subclass--that would centralize the read/write logic and a little traversal. `VisitProjectDependencies` assumes you're treating the `JObject` tree as mutable, though, and this code is treating it as immutable. I'm a fan of immutability, but I chose not to do that in the validation/upgrade logic because without the right structure it (imo) makes things more verbose.